sdk/src/main/scala/io/horizen/account/state/McAddrOwnershipMsgProcessor.scala

```diff
@@ -0,0 +1,370 @@
+    val v_barr = getUnsignedByteArray(mcSignature.getV)
+    val r_barr = padWithZeroBytes(getUnsignedByteArray(mcSignature.getR), SIGNATURE_RS_SIZE)
+    val s_barr = padWithZeroBytes(getUnsignedByteArray(mcSignature.getS), SIGNATURE_RS_SIZE)
+
+    val signatureData = new Sign.SignatureData(v_barr, r_barr, s_barr)
+
+    val hashedMsg = getMcHashedMsg(BytesUtils.toHexString(scAddress.pubKeyBytes()))
+
+    // verify MC message signature
+    val recPubKey = Sign.signedMessageHashToKey(hashedMsg, signatureData)
+    val recUncompressedPubKeyBytes = Bytes.concat(Array[Byte](0x04), Numeric.toBytesPadded(recPubKey, PUBLIC_KEY_SIZE))
+    val ecpointRec = ecParameters.getCurve.decodePoint(recUncompressedPubKeyBytes)
+    val recCompressedPubKeyBytes = ecpointRec.getEncoded(true)
+    val mcPubkeyhash = Ripemd160Sha256Hash(recCompressedPubKeyBytes)
+    val computedTaddr = toHorizenPublicKeyAddress(mcPubkeyhash, params)
+
+    computedTaddr.equals(mcTransparentAddress)
+
+  }
+
+  // this reproduces the MC way of getting a message for signing it via rpc signmessage cmd
+  private def getMcHashedMsg(messageToSignString: String) = {
+    // this is the magic string prepended in zend to the message to be signed*/
+    val strMessageMagic = "Zcash Signed Message:\n"
+    // compute the message to be signed. Similarly to what MC does, we must prepend the size of the byte buffers
+    // we are using
+    val messageMagicBytes = strMessageMagic.getBytes(StandardCharsets.UTF_8)
+    val mmb2 = Bytes.concat(Array[Byte](messageMagicBytes.length.asInstanceOf[Byte]), messageMagicBytes)
+
+    // TODO: currently size < 256 which is ok for a sc address; make it generic with int_to_bytes
+    val messageToSignBytes = messageToSignString.getBytes(StandardCharsets.UTF_8)
+    val mts2 = Bytes.concat(Array[Byte](messageToSignBytes.length.asInstanceOf[Byte]), messageToSignBytes)
+    // hash the message as MC does (double sha256)
+    doubleSHA256Hash(Bytes.concat(mmb2, mts2))
+  }
+
+  def doAddNewOwnershipCmd(msg: Message, view: BaseAccountStateView): Array[Byte] = {
+
+    // check that message contains a nonce, in the context of RPC calls the nonce might be missing
+    if (msg.getNonce == null) {
+      throw new ExecutionRevertedException("Call must include a nonce")
+    }
+
+    // check that msg.value is zero
+    if (msg.getValue.signum() != 0) {
+      throw new ExecutionRevertedException("Value must not be zero")
+    }
+
+    // check that sender account exists
+    if (!view.accountExists(msg.getFrom) ) {
+      throw new ExecutionRevertedException(s"Sender account does not exist: ${msg.getFrom}")
+    }
+
+    val inputParams = getArgumentsFromData(msg.getData)
+
+    val cmdInput = AddNewOwnershipCmdInputDecoder.decode(inputParams)
+    val scAddress = new AddressProposition(cmdInput.scAddress)
+    val mcTransparentAddress = cmdInput.mcTransparentAddress
+    val mcSignature = cmdInput.mcSignature
+
+    if (!msg.getFrom.equals(cmdInput.scAddress)) {
```

Comment:
I noticed that the **scAdress** is enforced to be equal to the **msg.getFrom** in both **addNewOwnership** and **removeOwnership**, does it make sense to remove this additional field (scAddress)?

Reply:
You are right, I removed it